FTN/ring_buffer.c

```diff
@@ -0,0 +1,167 @@
+    return 1;
+}
+bool RB_extract(ring_buffer *ring_buff, msg *out_msg)
+{
+    int64_t index_to_read = 0;
+    int actual_count = __atomic_sub_fetch(&ring_buff->count, 1, __ATOMIC_SEQ_CST);
+    if (actual_count < 0)
+    {
+        __atomic_add_fetch(&ring_buff->count, 1, __ATOMIC_SEQ_CST);
+        return 0;
+    }
+    index_to_read = __atomic_fetch_add(&(ring_buff->head_in_prograss), 1, __ATOMIC_SEQ_CST);
+    memcpy(out_msg, &ring_buff->msgs[(index_to_read % RING_BUFFER_SIZE)], sizeof(msg));
+    while ((ring_buff->head) % RING_BUFFER_SIZE != (index_to_read % RING_BUFFER_SIZE))
+    {
+    }
+    __atomic_fetch_add(&(ring_buff->head), 1, __ATOMIC_SEQ_CST);
+    __atomic_sub_fetch(&(ring_buff->len_in_prograss), 1, __ATOMIC_SEQ_CST);
+    return 1;
+}
+
+/* LESS protected version of insert/extract
+bool RB_insert(ring_buffer *ring_buff, msg *new_msg)
+{
+    if (!ring_buff || !new_msg)
+    {
+        return false;
+    }
+    if (RB_is_full(ring_buff))
+    {
+        printf("Queue is Full, couldnt write\n");
+        return false;
+    }
+
+    memcpy(&ring_buff->msgs[ring_buff->tail], new_msg, sizeof(msg));
+    if (CLIENTS[GET_PRIVATE_ID].id == 1){
+        printf("count: %ld <------------------: head: %ld, tailg got insert: %ld\n", ring_buff->count + 1, ring_buff->head, ring_buff->tail);
+        DumpHex((ring_buff->msgs[ring_buff->tail]).msg, (ring_buff->msgs[ring_buff->tail]).len); // <<====
+        printf(" <----------------------------------\n");
+    }
+    ring_buff->tail++;
+
+    if (ring_buff->tail == RING_BUFFER_SIZE) // end of circular buffer?
+    {
+        ring_buff->tail = 0; // turn around
+    }
+
+    ring_buff->count++;
+
+    return true;
+}
+
+bool RB_extract(ring_buffer *ring_buff, msg *out_msg)
+{
+    if (RB_empty(ring_buff))
+    {
+        printf("RB_extract- Queue is Empty\n");
+        return false;
+    }
+    memcpy(out_msg, &ring_buff->msgs[ring_buff->head], sizeof(msg));
+
+    printf("%ld ------------------>: head of extract: %ld, tail: %ld\n", e++, ring_buff->head, ring_buff->tail);
+    DumpHex((ring_buff->msgs[ring_buff->head]).msg, (ring_buff->msgs[ring_buff->head]).len); // <<====
+    printf(" ---------------------------------->\n");
+
+    ring_buff->head++;
+
+    printf("thread couses: RB_extract: tail- %ld, head- %ld, len of stored data: %ld, seq_num: %ld, total in this cli: %ld\n", ring_buff->tail, ring_buff->head, ring_buff->msgs[ring_buff->tail].len, ring_buff->msgs[ring_buff->tail].seq_num, ring_buff->count + 1);
+    if (ring_buff->head == RING_BUFFER_SIZE) // end of circular buffer?
+    {
+        printf("***********\n");
+        ring_buff->head = 0; // turn around
+    }
+
+    ring_buff->count--;
+    return true;
+}
+*/
+bool RB_peek(ring_buffer *ring_buff, uint32_t *out_index)
+{
+    if (RB_empty(ring_buff))
+    {
+        // printf("RB_peek- Queue is Empty\n");
+        return false;
+    }
+
+    *out_index = ring_buff->head % RING_BUFFER_SIZE;
```

Comment:
BUG! time of check time of use on empty.
what if the ring_buff is not empty and until we reach this live, somebody pop the ring_buffer?